racket/collects/racket/file.rkt

```diff
@@ -134,60 +135,112 @@
     ;; Do nothing with an immediately relative path or a root directory
     (void)]))
 
+(define-for-syntax (infer-temporary-file-template stx)
+  (define line   (syntax-line stx))
+  (define col    (syntax-column stx))
+  (define source (syntax-source stx))
+  (define pos    (syntax-position stx))
+  (define str-src
+    (cond [(path? source)
+           (regexp-replace #rx"^<(.*?)>(?=/)"
+                           (path->relative-string/library source)
+                           (lambda (_ s) (string-upcase s)))]
+          [(string? source) source]
+          [else #f]))
+  (define str-loc
+    (cond [(and line col) (format "-~a-~a" line col)]
+          [pos (format "--~a" pos)]
+          [else ""]))
+  (define combined-str (string-append (or str-src "rkttmp") str-loc))
+  (define sanitized-str (regexp-replace* #rx"[<>:\"/\\|]" combined-str "-"))
+  (define max-len 50) ;; must be even
+  (define not-too-long-str
+    (cond [(< max-len (string-length sanitized-str))
+           (string-append (substring sanitized-str 0 (- (/ max-len 2) 2))
+                          "----"
+                          (substring sanitized-str
+                                     (- (string-length sanitized-str)
+                                        (- (/ max-len 2) 2))))]
+          [else sanitized-str]))
+  (string-append not-too-long-str "_~a"))
+
+(define-syntax (make-temporary-directory stx)
+  (with-syntax ([app (datum->syntax stx #'#%app stx)])
+    (syntax-case stx ()
+      [x
+       (identifier? #'x)
+       #'make-temporary-directory/proc]
+      [(_)
+       #`(app make-temporary-directory/proc
+              '#,(infer-temporary-file-template stx))]
+      [(_ #:base-dir base-dir)
+       (not (keyword? #'base-dir))
+       #`(app make-temporary-directory/proc
+              '#,(infer-temporary-file-template stx)
+              #:base-dir #'base-dir)]
```

Comment:
The current `make-temporary-file` takes care to use `#%app` with the context of the source syntax, but I'm not sure I've fully considered any implications for using keyword arguments.